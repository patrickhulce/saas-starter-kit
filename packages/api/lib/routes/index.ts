import {IRouterMap} from 'klay'

import {accountsRouterOptions} from './accounts'
import {oauthRouterOptions} from './oauth'
import {usersRouterOptions} from './users'

export const routerMap: IRouterMap = {
  '/oauth': oauthRouterOptions,
  '/accounts': accountsRouterOptions,
  '/users': usersRouterOptions,
}
