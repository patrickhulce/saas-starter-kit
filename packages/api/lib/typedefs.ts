import 'express' // tslint:disable-line

declare module 'express-serve-static-core' {
  /* tslint:disable */
  export interface Response {
    body?: any
  }
  /* tslint:enable */
}

export {}
