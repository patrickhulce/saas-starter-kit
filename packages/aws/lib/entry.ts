process.env.APP_ENV = 'firebase-production'
export * from './http'
