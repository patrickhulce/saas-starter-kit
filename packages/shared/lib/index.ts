export {default as conf} from './conf'
export * from './model-context'
export * from './kiln'
export * from './typedefs'
export * from './logger'
export {configuration as authConfiguration} from './auth'
