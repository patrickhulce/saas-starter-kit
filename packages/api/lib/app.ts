import {json} from 'body-parser'
import * as cookies from 'cookie-parser'
import * as express from 'express'
import {createAndMergeRouters, createGrantCreationMiddleware} from 'klay'
import {omit, pick} from 'lodash'

import {authConfiguration, kiln, sqlExtension} from '../../shared/lib'

import {setPrivateCacheControl} from './middleware/cache-control'
import {handleError} from './middleware/handle-error'
import {handlePageRequest, pageRoutes} from './middleware/handle-page-request'
import {handlePromise} from './middleware/handle-promise'
import {loggers} from './middleware/logging'
import {routerMap} from './routes'
import './typedefs' // tslint:disable-line

const app: express.Express = express()

app.use(json({strict: false}))
app.use(loggers)
app.use(cookies())
app.use(setPrivateCacheControl)
app.use(createGrantCreationMiddleware(authConfiguration))

app.get(pageRoutes, handlePageRequest)
app.use('/api/v1', createAndMergeRouters(kiln, routerMap).router)

app.use(handlePromise)
app.use(handleError)

export {app, sqlExtension}
