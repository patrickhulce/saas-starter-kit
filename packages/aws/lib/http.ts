import * as awsServerlessExpress from 'aws-serverless-express'
import {app} from '../../api/lib/app'

const server = awsServerlessExpress.createServer(app)

export function handler(event: any, context: any) {
  return awsServerlessExpress.proxy(server, event, context)
}
