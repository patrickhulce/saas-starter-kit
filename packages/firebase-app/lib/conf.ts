import * as functions from 'firebase-functions'
import {IConf} from '../../shared/lib/typedefs'
import conf from '../../shared/lib/conf'

conf.production = !!functions.config().production
conf.apiPathPrefix = '/api'
conf.database.connectionURL = functions.config().sqlURL

export default conf
