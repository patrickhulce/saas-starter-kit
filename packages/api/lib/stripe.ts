import * as Stripe from 'stripe'

import {accountExecutor} from '../../shared/lib'
import conf from '../../shared/lib/conf'
import {AccountPlan, IUser} from '../../shared/lib/typedefs'

// TODO: add stripe webhooks

export const stripe = new Stripe(conf.stripe.secretKey)

export async function createOrUpdateStripe(
  user: IUser,
  plan: AccountPlan,
  stripeSourceId: string,
): Promise<void> {
  const account = await accountExecutor.findByIdOrThrow(user.accountId)
  // TODO: update payment information too
  if (account.stripeId) {
    throw new Error('Update not yet supported')
  }

  const customer = await createStripeCustomer(user, plan, stripeSourceId)
  account.stripeId = customer.id
  // TODO: poll stripe until charge goes through
  account.plan = plan
  await accountExecutor.update(account)
}

async function createStripeCustomer(
  user: IUser,
  plan: AccountPlan,
  stripeSourceId: string,
): Promise<Stripe.customers.ICustomer> {
  return stripe.customers.create({
    email: user.email,
    description: `${user.firstName} ${user.lastName}`,
    metadata: {accountId: user.accountId, userId: user.id!},
    source: stripeSourceId,
    plan,
  })
}
