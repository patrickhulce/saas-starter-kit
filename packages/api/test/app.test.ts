import {app, sqlExtension} from '../lib/app'

describe('lib/app.ts', () => {
  let server

  it('should export a working server', done => {
    server = app.listen(done)
  })

  it('should disconnect from SQL gracefully', async () => {
    await sqlExtension.sequelize.close()
  })

  it('should disconnect express gracefully', done => {
    server.close(done)
  })
})
