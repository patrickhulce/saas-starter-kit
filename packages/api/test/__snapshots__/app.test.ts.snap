// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/app.ts account setup should have sent a welcome email 1`] = `
Object {
  "content": Object {
    "from": "noreply@THE_DOMAIN",
    "html": "Hello Admin User,<br><br>You're receiving this message because you registered an account for THE_PRODUCT_DISPLAY_NAME.<br>Please follow the link below to verify your email address.<br><br><a href=\\"http://localhost/api/v1/users/verifications?key=<key>\\">http://localhost/api/v1/users/verifications?key=<key></a><br><br>If you did not register an account with us, you can ignore this email.<br><br>Thanks,<br>The THE_PRODUCT_DISPLAY_NAME Team",
    "subject": "Welcome to THE_PRODUCT_DISPLAY_NAME! - Verify your email",
    "text": "Hello Admin User,

You're receiving this message because you registered an account for THE_PRODUCT_DISPLAY_NAME.
Please follow the link below to verify your email address.

http://localhost/api/v1/users/verifications?key=<key>

If you did not register an account with us, you can ignore this email.

Thanks,
The THE_PRODUCT_DISPLAY_NAME Team",
  },
  "options": Object {
    "transactional": true,
  },
  "recipients": Array [
    Object {
      "address": Object {
        "email": "<uuid>@example.com.sink.sparkpostmail.com",
        "name": "Admin User",
      },
      "tags": Array [
        "welcome",
        "email-verification",
      ],
    },
  ],
}
`;
